Wod3

TheTisiboth opened the pull request:
All the different features are references by the Issues ID in the commit message

WesIngwersen commented:
@TheTisiboth @LorenHeyns  This code is updated for v1.1 of the USEEIO_API, but the USEEIO_API on the production server has not been updated yet ..it's still at v1.0.  The main issue is that this version is trying to call the `N` matrix but this doesn't yet exist on the production API - it is still called the `U` matrix in that version.  So  this code in this branch needs to be reverted back to use U in order to test and (if passing otherwise), merge it.

WesIngwersen commented:
This is a JS error i'm getting related to my previous comment when testing the local Sector List Alone widgets - it reads like a CORS error but the problem is that N.json does not exist
```
Access to XMLHttpRequest at 'file:///C:/Users/wesle/useeio-widgets-tisiboth/build/api/USEEIOv2.0/matrix/N.json' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, chrome-untrusted, https.
api/USEEIOv2.0/matrix/N.json:1 Failed to load resource: net::ERR_FAILED
```

WesIngwersen commented:
@TheTisiboth . Pinging you here. These issue with the N matrix reverting back to U was never addressed, so this has not been merged yet,

TheTisiboth commented:
Hello @WesIngwersen. Sorry for the late answer, and this unsolved issue, but I just come back from vacation. 
I am going to roll back and use the USEEIO model v1.2, so it uses the U matrix

TheTisiboth commented:
It is now using the USEEIOv1.2 version. You can however switch back to the v2.0 in the future, and it will automatically use the N matrix, instead of the U one

TheTisiboth commented:
> 
> 
> @LorenHeyns @TheTisiboth I don't see the value with including the description of the 3 types of demand as you break them down here on this page.
> ![image](https://user-images.githubusercontent.com/5831306/131161800-58bab271-16a2-4b85-b8a4-2f3c4b3f806c.png)

@LorenHeyns wanted to have an explanation to explain the differences between a null value and a 0 value for the demand. What would you rather recommend ?

msrocka commented on src/webapi.ts:
```diff
@@ -649,6 +660,19 @@ export class Model {
+        // So we check the model version, to decide wether we use the N or the U matrix
+        if (name === "N") {
+          try {
+            const modelVersion = parseInt(this._conf.model.split("v")[1]);
```
This is maybe a working fix for now but in general we cannot assume that the model ID always follows a pattern `[prefix]v[major].[minor]`, or can we? Better maybe just match with a regex `*v1.*` without throwing errors here?

TheTisiboth replied:
You are right, I was assuming that the model ID had a predefined pattern.
However, I should throw an error if I can't find the version number, since it is used for knowing which matrix we should use (N or U), isn't it ?
Anyway, I am using the following pattern now `/v(\d*\.?\d*)/i`, to find the version number (it looks for a v followed by a number)

WesIngwersen replied:
versioning will be like 
`v1.2`
`v2.0.1` 
so a wildcard is appropriate

msrocka commented on src/webapi.ts:
```diff
@@ -676,7 +700,7 @@ export class Model {
      * data.
      */
-    async calculate(setup: CalculationSetup): Promise<Result> {
+    async calculate(setup: CalculationSetup, scaleFactor = 1000000): Promise<Result> {
```
Why would applying such a scaling factor here be correct? The demand vector should scale the result so it would be better to provide a better demand vector here?

WesIngwersen replied:
@msrocka @TheTisiboth 
I agree that you can not apply a scaling factor to the calculate function.  I understood for the input-output chart that were just scaling the `D` matrix. This is simply to show per `x`  (e.g. million) dollar what the values were, instead of per dollar.

msrocka commented on src/widgets/iogrid/commodity-list.tsx:
```diff
@@ -540,3 +629,22 @@ const CheckBox = (props: { checked: boolean }) =>
+        return r.toFixed(decimal);
+
+    let n;
+    do {
```
Maybe better something like:

```ts
function formatNumber(x: number): string {
    if (!x)
        return "0.000";
    if (x > 1)
        return x.toFixed(3);
    const digits = Math.max(3, Math.ceil(Math.log10(1/x)))
    return x.toFixed(digits);
}
```

TheTisiboth replied:
Oh yeah thank you for this, it is way more efficient!

msrocka commented on src/widgets/iogrid/flow-list.tsx:
```diff
@@ -196,3 +201,22 @@ export const FlowList = (props: {
+/**
+* Increases the number of decimal digits until the number has the right number of digits
+*/
+function formatNumber(r: number) {
```
This is the same format function as in another module. Would be good to extract this into a utility module and also replace this with a more efficient method (see the other comment).

msrocka commented on src/widgets/sector-list/impacts.tsx:
```diff
@@ -194,3 +234,56 @@ export const ImpactResult = (props: RowProps) => {
+/**
+ * Increases the number of decimal digits until the number has the right number of digits
+ */
+function formatDecimalNumber(r: number) {
```
again the format function...

msrocka commented on src/widgets/sector-list/sector-list.tsx:
```diff
@@ -166,11 +161,18 @@ export class SectorList extends Widget {
 }
 
 async function calculate(model: Model, config: Config): Promise<HeatmapResult> {
+    const scaleFactor = config.scale_factor ? config.scale_factor : 1000000;
```
I assume the scaling factor was introduced here because no demand vector was selected via the settings, right? Not sure with this...

msrocka reviewed (approved):
I am not sure with applying a scaling factor on the /1USD-result matrices. For me these results should be scaled by a demand vector. It was done before when the respective analysis settings were provided via the URL parameters. But if this is how it should work I am fine with this… The number formatting could be done in a utility method so that this can be optimized in a single place. Also, it breaks when the model ID does not follow a pattern `[prefix]v[major].[minor]`, right? But for me this can be merged.

TheTisiboth commented:
> 
> 
> I get this on running npm install - not sure if it's an issue
> 
> ```
> npm WARN useeio-widgets@0.1.0 scripts['server'] should probably be scripts['start'].
> ```

No, it is just a warning, it is not important, it is just about naming convention

TheTisiboth commented:
> 
> 
> iotables_config.html When I select the [Sector List with Inflow and Outflow - Rubber tire manufacturing example](http://127.0.0.1:8080/iotables.html#sectors=326210&page=5), I see the sector named in the URL but I don't see it selected - its not present in the view but I would think all selected sectors should be at the top

I changed the behavior: 
If we specify as URL parameters some indicators, then the rows are sorted according to those indicators.
Otherwise, if no indicators are provided, then the rows are sorted, with the selected rows on top. Does it looks good to you ?

modelearth commented:
I'm seeing two different lists, depending on how I reload the following page.

https://thetisiboth.github.io/useeio-widget-builds/iotables.html#sectors=326210&page=1

If I open a new tab, paste the link and hit return, I don't see 326210 at the top.

I also don't see it at the top here:

https://thetisiboth.github.io/useeio-widget-builds/iotables.html#sectors=326210

If I then append &page=1 to the link above and hit return, I finally see Rubber Tires at the top.
If I hit reload, it goes away.

TheTisiboth commented:
> iotables.html#sectors=326210

It is indeed a bug, which is already fixed. Should I however remove this widget ?

modelearth commented:
Let's leave the old sample in there, but comment out all but one link to it.
Call that remaining link: Version without Google Material Design (or whatever is appropriate)
Change existing links to use the newer version for Tires, etc.

Since the older version is a non-material-design layout, it may be useful for stand-alone lists in the future.

I like the grey color myself rather than the red bars. You could add this as a new issue:
Show red when in the top 33.3% adverse, grey when in the middle, and green when positive 33.3%.
We're going to carry those colors over to a "nutrition" label currently in the works:
https://model.earth/io/template/

I also like how the older version aligns the text against the side margins.
Later we can try modifying the material design css.

Please place a settings widget above one of the new version pages, if we don't have that already.

TheTisiboth commented:
@modelearth I updated the IOChart as requested. By the way, I also updated to the latest version of Material UI (and react), so the IOChart looks a bit nicer.
Is there anything else to modify/add ?

WesIngwersen reviewed (changes requested):
I changed my review status. I think we still need to resolve the scaling issue.  We might need to meet about this to discuss because I'm not sure myself where all this is used. 
@TheTisiboth @msrocka @LorenHeyns

TheTisiboth commented:
I removed the scale factor, and addes some widget description (which is really short for now).
If you agree with this, we can merge.

modelearth commented:
Looks good to merge. Responsive columns are working. Thanks!
